datalad/distribution/dataset.py

```diff
@@ -102,14 +103,44 @@ def repo(self):
                             pass
                 if self._repo is None:
                     lgr.info("Failed to detect a valid repo at %s" % self.path)
+                else:
+                    # reset ID, because we now have a repo attached
+                    self._id = None
         elif not isinstance(self._repo, AnnexRepo):
             # repo was initially set to be self._repo but might become AnnexRepo
             # at a later moment, so check if it didn't happen
             if 'git-annex' in self._repo.get_branches():
                 # we acquired git-annex branch
                 self._repo = AnnexRepo(self._repo.path, create=False)
+                # reset ID, because we now have an annex attached
+                self._id = None
```

Comment:
So it would possibly change in the lifetime of a dataset... What is the purpose of id here since we wouldn't then rely on id to correspond to the dataset? ( Since it might change)